server/lib/currency.ts

```diff
@@ -264,3 +265,20 @@ export function reduceArrayToCurrency(array: AmountWithCurrencyAndDate[], curren
     },
   );
 }
+
+/**
+ * Gets exchange rate between given crypto currency and fiat.
+ */
+export async function getCryptoFxRate(cryptoCurrency: string, fiatCurrency: string): Promise<string> {
+  if (!cryptoCurrency || !fiatCurrency) {
+    return null;
+  }
+  try {
+    const { rate: cryptoToUSDRate } = await getCryptoToUSDRate(cryptoCurrency);
+    const USDToCollectiveCurrencyRate = await getFxRate('USD', fiatCurrency);
+    return cryptoToUSDRate * USDToCollectiveCurrencyRate;
+  } catch (error) {
+    logger.error(error);
+    return null;
+  }
+}
```

Comment:
@znarf : This is interesting. I spent some time looking into how line 144 of `server/paymentProviders/thegivingblock/index.js` and I think there's a problem there. It uses the `fetchFxRates` function internally. And that in turn uses, `https://data.fixer.io/`. I think this might work for basic crypto like `BTC` but do not support other crypto currencies. If we are going this route, we have to use, https://coinlayer.com/. 

The complete list of what `data.fixer.io` supports is given at by `http://data.fixer.io/api/latest?access_key = API_KEY` endpoint. And this is what I got; 

```
{"AED":3.843287,"AFN":94.698551,"ALL":119.597048,"AMD":478.695814,"ANG":1.886015,"AOA":429.881799,"ARS":123.461607,"AUD":1.504471,"AWG":1.883956,"AZN":1.777998,"BAM":1.945875,"BBD":2.11297,"BDT":91.61939,"BGN":1.955595,"BHD":0.39441,"BIF":2106.306088,"BMD":1.046352,"BND":1.451604,"BOB":7.205001,"BRL":5.2004,"BSD":1.046436,"BTC":3.5896715e-5,"BTN":81.109262,"BWP":12.730946,"BYN":3.531532,"BYR":20508.494452,"BZD":2.109368,"CAD":1.349004,"CDF":2097.410525,"CHF":1.034631,"CLF":0.032412,"CLP":894.34848,"CNY":7.067587,"COP":4257.573912,"CRC":699.822946,"CUC":1.046352,"CUP":27.728322,"CVE":109.291137,"CZK":24.6668,"DJF":185.958022,"DKK":7.441811,"DOP":57.810521,"DZD":152.843753,"EGP":19.108161,"ERN":15.695277,"ETB":53.881779,"EUR":1,"FJD":2.286174,"FKP":0.856262,"GBP":0.847477,"GEL":3.102446,"GGP":0.856262,"GHS":8.004598,"GIP":0.856262,"GMD":56.660158,"GNF":9247.657522,"GTQ":8.026482,"GYD":218.935718,"HKD":8.213076,"HNL":25.582999,"HRK":7.535512,"HTG":116.681814,"HUF":383.791782,"IDR":15423.277227,"ILS":3.520519,"IMP":0.856262,"INR":81.431384,"IQD":1527.673566,"IRR":44260.67923,"ISK":139.32156,"JEP":0.856262,"JMD":161.968803,"JOD":0.741873,"JPY":134.161162,"KES":121.794106,"KGS":83.998504,"KHR":4248.188059,"KMF":492.256339,"KPW":941.716815,"KRW":1334.966689,"KWD":0.320704,"KYD":0.87203,"KZT":453.282298,"LAK":13843.234127,"LBP":1587.315328,"LKR":376.724339,"LRD":160.616623,"LSL":16.793831,"LTL":3.089605,"LVL":0.632928,"LYD":5.029125,"MAD":10.557161,"MDL":19.961774,"MGA":4231.968622,"MKD":61.30021,"MMK":1937.48614,"MNT":3218.338181,"MOP":8.45944,"MRO":373.547398,"MUR":45.251157,"MVR":16.156012,"MWK":852.776506,"MXN":20.984951,"MYR":4.599243,"MZN":66.788515,"NAD":16.794561,"NGN":434.424251,"NIO":37.386489,"NOK":10.360244,"NPR":129.769883,"NZD":1.661931,"OMR":0.40283,"PAB":1.046436,"PEN":3.949986,"PGK":3.677921,"PHP":54.907314,"PKR":206.971386,"PLN":4.647947,"PYG":7168.456934,"QAR":3.80979,"RON":4.947143,"RSD":117.484602,"RUB":68.532323,"RWF":1068.848321,"SAR":3.925027,"SBD":8.500065,"SCR":13.594645,"SDG":475.573167,"SEK":10.525984,"SGD":1.455501,"SHP":1.441242,"SLL":13550.255534,"SOS":606.345977,"SRD":22.055527,"STD":21657.368809,"SVC":9.155982,"SYP":2628.907015,"SZL":16.815016,"THB":36.266426,"TJS":13.070381,"TMT":3.672695,"TND":3.226894,"TOP":2.434913,"TRY":16.684392,"TTD":7.097819,"TWD":31.195717,"TZS":2433.813942,"UAH":30.917224,"UGX":3834.291762,"USD":1.046352,"UYU":43.339855,"UZS":11666.821997,"VEF":223741629001.3448,"VND":24213.626026,"VUV":119.562077,"WST":2.6989,"XAF":652.587114,"XAG":0.048847,"XAU":0.000576,"XCD":2.827818,"XDR":0.781465,"XOF":658.680865,"XPF":118.813571,"YER":261.907391,"ZAR":16.82596,"ZMK":9418.422908,"ZMW":17.81621,"ZWL":336.924839}
```

And trying to make a payment on staging using SUSHI, will print this error on the api; 

![image](https://user-images.githubusercontent.com/12435965/169165021-4ae297a7-4790-43c7-8e28-e0379cd025be.png)

So I think even there we have to use the same approach of using the currency returned by TGB as I've done in this PR. WDYT? If you agree, I will open a new PR and work on this next. 🤔

Reply:
The code is designed for failure and it's not a problem:

```
 // update approximate amount in order currency
  let cryptoFxRate;
  try {
    cryptoFxRate = await getFxRate(order.data.thegivingblock.pledgeCurrency, order.currency);
  } catch (e) {
    console.log(e);
  }
  if (cryptoFxRate) {
    const totalAmount = Math.round(order.data.thegivingblock.pledgeAmount * cryptoFxRate * 100);
    await order.update({ totalAmount });
  }
```

What are the crypto currencies we're currently supporting and Fixer not? Should we care?

Reply:
Yes, the failure isn't necessarily a problem, since even if the crypto currency exchange rate doesn't exist on fixer that doesn't matter as this just for the interim before the transaction is verified. If the user actually contributes this amount then using we will [update that amount to what's returned from TGB](https://github.com/opencollective/opencollective-api/blob/869a5fb0c9e6144c57be4c219602be3183f2647e/server/paymentProviders/thegivingblock/webhook.js#L37). 

So I don't think we should care about this; however when showing the exchange rate on the we cannot use `fetchFxRates` as that doesn't return the correct amount. So your initial question of, 

> At line 144 of `server/paymentProviders/thegivingblock/index.js` we're fetching a crypto fxRate without the need of this. What is justifying it?

we have to use a method to obtain the exchange rate that returns the correct values for all supported cryptocurrencies of TGB. This is the reason I used the specialized function with a call to the TGB endpoint; `crypto-to-usd-rate`

Reply:
> What are the crypto currencies we're currently supporting and Fixer not? 

Fixer seems to support `BTC` only. If we want more using their API we should use https://coinlayer.com/